### PR TITLE
GODRIVER-2603 (Contd.) Revised error handling using Go 1.13 error APIs

### PR DESCRIPTION
**Pull request by kumarlokesh**
[GODRIVER-2603](https://jira.mongodb.org/browse/GODRIVER-2603)

## Summary
Follow-up PR to https://github.com/mongodb/mongo-go-driver/pull/1470 and https://github.com/mongodb/mongo-go-driver/pull/1476

## Background & Motivation
As a requirement for the related ticket [GODRIVER-2646](https://jira.mongodb.org/browse/GODRIVER-2646), all errors are required to
be addressed to enable [errorlint linter](https://github.com/mongodb/mongo-go-driver/blob/0a826e6efd4b362d59be309e2cd9e6f70ddcc5c0/.golangci.yml#L16).

Would submit subsequent PR(s) to fix remaining errors before `errorlint` linter
could be enabled.

### CODE REVIEW

File: mongo/errors.go [hunk 1]
```diff
@@ -52,10 +52,12 @@ func replaceErrors(err error) error {
...
 	}
 
-	if err == topology.ErrTopologyClosed {
+	if errors.Is(err, topology.ErrTopologyClosed) {
 		return ErrClientDisconnected
 	}
-	if de, ok := err.(driver.Error); ok {
+
+	var de driver.Error
+	if errors.As(err, &de) {
```
Comment (prestonvasquez) [inline mongo/errors.go:60]: This comparison is not the same as `de, ok := err.(driver.Error); ok` in that *driver.Error does not implement the [error](https://pkg.go.dev/builtin#error) interface. I'm realizing that this pattern of using `errors.As` with a to compare to a target who's sole purpose is the comparison. I think we should avoid doing this and retain the existing type assertion.

Would you mind revert all code with this pattern:

```go
var target Error
if errors.As(err, &target) {}
```

Back to using a type assertion?

**Review by prestonvasquez — CHANGES_REQUESTED** [review]
Thanks for the additional work on error handling @kumarlokesh !

File: mongo/errors.go (hunk 1 above)
Comment (kumarlokesh) [reply mongo/errors.go]: Thanks @prestonvasquez for the review and sharing feedback!

Updated PR.

**Comment by blink1073** [comment]
@prestonvasquez I tried the automatic backport but it looks like there are conflicts from https://github.com/mongodb/mongo-go-driver/pull/1486 that I didn't know how to resolve myself.  

You can run:

```bash
git pull upstream v1
git create -b GODRIVER-2603 upstream/v1
git cherry-pick -x 4dbe540
```

And then handle the merge conflicts.

**Pull request merged**